### ISSUE TITLE
Expose ErrorDescription for CA failures

### ISSUE THREAD

**Issue by nazukj** [AzureAD/azure-activedirectory-library-for-android#1006]
From Brandon Zahl: 
I'm driving a feature on Office called Links Open Right (LOR) and LOR needs to be aware of conditional access. Now when conditional access takes place, the clients are not allowed to access a document. From talking with Conditional Access folks, we decided that the best thing is to detect the open failed due to conditional access policy and then launch the URL in the browser. The user may actually have access to the document thru the web browser and if not, they will see a more informative error message explaining the CA failure.
When I met with CA folks to discuss solution we decided on a short-term which involved parsing error description.

Ask is for Android is to expose the error description as a short term mitigation.

**Comment by heidijinxujia** [comment]
Quote an example from Remy's email.
Here’s an example error, returned in a 400 response from https://login.windows-ppe.net/common/oauth2/token when the token request failed due to a conditional access policy:
 
{"error":"interaction_required","error_description":"AADSTS50131: Windows device is not in required device state: domain_joined\r\nTrace ID: 41f28fdd-8c37-4f3e-8e51-2eeea28e0700\r\nCorrelation ID: 058ff775-5dfe-4e0a-901f-2440103cfb59\r\nTimestamp: 2017-10-11 00:20:23Z","error_codes":[50131],"timestamp":"2017-10-11 00:20:23Z","trace_id":"41f28fdd-8c37-4f3e-8e51-2eeea28e0700","correlation_id":"058ff775-5dfe-4e0a-901f-2440103cfb59"}

In ADAL Android, when receiving HttpURLConnection.HTTP_BAD_REQUEST,  parseJsonResponse()
 will detect the "error" key from the response and reply AuthenticationResult with corresponding OAuth2 error code and error description. (Ref: com.microsoft.aad.adal.Oauth2#processUIResponseParamsL#252)
```java
if (response.containsKey(AuthenticationConstants.OAuth2.ERROR)) {
    ...
    result = new AuthenticationResult(response.get(AuthenticationConstants.OAuth2.ERROR),
    response.get(AuthenticationConstants.OAuth2.ERROR_DESCRIPTION),
    response.get(AuthenticationConstants.OAuth2.ERROR_CODES));
}
```
And these  error_codes and error_description will be wrapped and passed into message para of AuthenticationException with ADALError.Auth_FAILED to the calling app.
```java
        if (!StringExtensions.isNullOrBlank(result.getErrorCode())) {
            Logger.e(TAG, result.getErrorLogInfo(), null, ADALError.AUTH_FAILED);
            throw new AuthenticationException(ADALError.AUTH_FAILED,
                    result.getErrorLogInfo());
        }
```

Currently, if the client app can get the error_description and error_code by calling AuthenticationException.getMessage(). 

A long-term fix will be combine the fix for #1046 and create new exception with status_code, headers and error description as properties.

**Comment by heidijinxujia** [comment]
Emailed the customer with instructions.
